monai/transforms/spatial/array.py

```diff
@@ -82,12 +85,167 @@
+        dst = to_affine_nd(spatial_rank, dst) if dst is not None else src
+        dst, *_ = convert_to_dst_type(dst, dst, dtype=torch.float32)
+
+        if np.allclose(src, dst, atol=AFFINE_TOL):
+            # no significant change, return original image
+            output_data, *_ = convert_to_dst_type(img, img, dtype=torch.float32)
+            return output_data, dst
+
+        if has_nib and isinstance(img, np.ndarray):
+            spatial_ornt, dst_r = reorient_spatial_axes(img.shape[1 : spatial_rank + 1], src, dst)
+            if np.allclose(dst_r, dst, atol=AFFINE_TOL):
+                # simple reorientation achieves the desired affine
+                spatial_ornt[:, 0] += 1
+                spatial_ornt = np.concatenate([np.array([[0, 1]]), spatial_ornt])
+                img_ = nib.orientations.apply_orientation(img, spatial_ornt)
+                output_data, *_ = convert_to_dst_type(img_, img, dtype=torch.float32)
+                return output_data, dst
+
+        xform = np.linalg.inv(src) @ dst
+        xform = to_affine_nd(spatial_rank, xform)
+        # no resampling if it's identity transform
+        if np.allclose(xform, np.diag(np.ones(len(xform))), atol=AFFINE_TOL):
+            output_data, *_ = convert_to_dst_type(img, img, dtype=torch.float32)
+            return output_data, dst
+
+        _dtype = dtype or self.dtype or img.dtype
+        spatial_size = ensure_tuple(spatial_size)
+        in_spatial_size = list(img.shape[1 : spatial_rank + 1])
+        if spatial_size[0] == -1:  # if the spatial_size == -1
+            spatial_size = in_spatial_size
+        elif spatial_size[0] is None:
+            spatial_size, _ = compute_shape_offset(in_spatial_size, src, dst)  # type: ignore
+        spatial_size = spatial_size[:spatial_rank]
+        chns, additional_dims = img.shape[0], img.shape[spatial_rank + 1 :]  # beyond three spatial dims
+        # resample
+        img_ = convert_data_type(img, torch.Tensor, dtype=_dtype)[0]  # type: ignore
+        xform = convert_to_dst_type(xform, img_)[0]  # type: ignore
+        align_corners = self.align_corners if align_corners is None else align_corners
+        mode = look_up_option(mode or self.mode, GridSampleMode)
+        padding_mode = look_up_option(padding_mode or self.padding_mode, GridSamplePadMode)
+        if additional_dims:
+            xform_shape = [-1] + in_spatial_size
+            img_ = img_.reshape(xform_shape)
+        if align_corners:
+            _t_r = torch.diag(torch.ones(len(xform), dtype=xform.dtype, device=xform.device))  # type: ignore
+            for idx, d_dst in enumerate(spatial_size[:spatial_rank]):
+                _t_r[idx, -1] = (max(d_dst, 2) - 1.0) / 2.0
+            xform = xform @ _t_r
+            if not USE_COMPILED:
+                _t_l = normalize_transform(in_spatial_size, xform.device, xform.dtype, align_corners=True)  # type: ignore
+                xform = _t_l @ xform
+            affine_xform = Affine(
```

Comment:
Object creation is going to be a bit expensive, though this object doesn't create large data segments, so trying to use this transform quickly for a large number of smallish images is going to suffer for a constant runtime cost.

Reply:
agree with the idea in general, but in this case the `affine` has to be provided at the constructor of `Affine`, because the grid will be created within `__call__`.